sql: add index join and lookup join assertions

mgartner opened the pull request:
#### sql: add fetch row count assertions in vectorized index joiner

Assertions have been added to the vectorized index joiner that ensure
that an index join fetches the expected number of rows. If an index join
has a locking wait policy of `SKIP LOCKED`, it should fetch no more than
the number of input rows. Otherwise, an index join should fetch exactly
one row for each input row. If these assertions fail, the query results
may be incorrect, e.g., due to index corruption, and an internal error
is preferred over a successful result.

Informs #135696

Release note: None

#### sql: add fetch row count assertions in row-by-row index joiner

Assertions have been added to the join reader that ensure that an index
join fetches the expected number of rows.

Informs #135696

Release note: None

#### sql: add fetch row count assertions in row-by-row lookup joiner

Assertions have been added to the join reader that ensure that a lookup
join on key columns fetches the expected number of rows.

Fixes #135696

Release note: None

mgartner committed:
sql: add fetch row count assertions in vectorized index joiner

Assertions have been added to the vectorized index joiner that ensure
that an index join fetches the expected number of rows. If an index join
has a locking wait policy of `SKIP LOCKED`, it should fetch no more than
the number of input rows. Otherwise, an index join should fetch exactly
one row for each input row. If these assertions fail, the query results
may be incorrect, e.g., due to index corruption, and an internal error
is preferred over a successful result.

Informs #135696

Release note: None

yuzefovich reviewed (commented):
Nice addition! I have a suggestion and also wonder whether for completeness we should do the same for `joinReader` with `indexJoinReaderType`.

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148538#-:-OT49tlIC-ZuZsNs2doR:b-77j1fk)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rytaft)

___
*[`pkg/sql/colfetcher/index_join.go` line 220 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148538#-OT48cDIBZ7ddHjFdb58:-OT48cDJ4rWqxgWAF3hy:bzh9128) ([raw file](https://github.com/cockroachdb/cockroach/blob/2faf7dc9d2fc4dacfc90001f2c77dff268d8dd87/pkg/sql/colfetcher/index_join.go#L220)):*
> ```Go
> 
> 			// For memory accounting, we assume the index join will return
> 			// exactly one output row per input row. This is this true most of
> ```

nit: "this is this true".
___
*[`pkg/sql/colfetcher/index_join.go` line 259 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148538#-OT494Aa3cm2dbcgDa57:-OT494Aa3cm2dbcgDa58:b-niajr8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2faf7dc9d2fc4dacfc90001f2c77dff268d8dd87/pkg/sql/colfetcher/index_join.go#L259)):*
> ```Go
> 				continue
> 			}
> 			if s.lockingWaitPolicy == descpb.ScanLockingWaitPolicy_SKIP_LOCKED {
> ```

In theory, there is no guarantee that `cFetcher` will return all looked up rows in a single `coldata.Batch` (e.g. if some rows have extremely large values, it might split them across multiple batches). ~(I think in practice it should happen given `setEstimatedRowCount` call before.)~ I think the correct way to do this assertion would be to count rows across all batches and do the assertion once we receive the first zero-length batch (which indicates that the cFetcher is exhausted), right before transitioning back to `indexJoinConstructingSpans` state.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148538)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[`pkg/sql/colfetcher/index_join.go` line 259 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148538#-OT494Aa3cm2dbcgDa57:-OT4H8x3BBHty6lcsTaq:bhxeyhm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2faf7dc9d2fc4dacfc90001f2c77dff268d8dd87/pkg/sql/colfetcher/index_join.go#L259)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

In theory, there is no guarantee that `cFetcher` will return all looked up rows in a single `coldata.Batch` (e.g. if some rows have extremely large values, it might split them across multiple batches). (I think in practice it should happen given `setEstimatedRowCount` call before.) I think the correct way to do this assertion would be to count rows across all batches and do the assertion once we receive the first zero-length batch (which indicates that the cFetcher is exhausted), right before transitioning back to `indexJoinConstructingSpans` state.
</blockquote></details>

Actually, we buffer input rows (i.e. spans) until the memory size, so we do expect to see multiple batches from the cFetcher, thus, we must do the check once the zero-length batch is received.


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/colfetcher/index_join.go` line 259 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148538#-OT494Aa3cm2dbcgDa57:-OT4Q5TQ4zcTX2zn3lAc:b-1dkoic) ([raw file](https://github.com/cockroachdb/cockroach/blob/2faf7dc9d2fc4dacfc90001f2c77dff268d8dd87/pkg/sql/colfetcher/index_join.go#L259)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Actually, we buffer input rows (i.e. spans) until the memory size, so we do expect to see multiple batches from the cFetcher, thus, we must do the check once the zero-length batch is received.
</blockquote></details>

This is super helpful context. Thank you! I'm glad a bunch of tests caught this too:

![image.png](https://files.reviewableusercontent.io/images/reviewable/1128750/Jod-knCCn2tdRkd_lwZBFyMp/image.png)


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> [I] also wonder whether for completeness we should do the same for `joinReader` with `indexJoinReaderType`.

Yes, that's the plan. I was hoping to get the index-join one done-and-dusted first. But since there are some problems here I need to fix, I might tack on a joiner reader commit too.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148538#-:-OT4QLL4D2RhGNEsCpyc:bw0nfb4)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rytaft and @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with a couple of nits. It seems reasonable to include the join reader assertions for both index and lookups joins, I'm ok merging this as is (given that the index joins should generally use the vectorized engine anyway).

> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148538#-:-OT5Lvf96YMeuW-D-4we:b-e2rd59)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[`pkg/sql/colfetcher/index_join.go` line 338 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148538#-OT5L_608KRzRm9OmxTJ:-OT5L_608KRzRm9OmxTK:b-b5ve3e) ([raw file](https://github.com/cockroachdb/cockroach/blob/058d27a26de242b914a4ca4da07059d8b66e4a46/pkg/sql/colfetcher/index_join.go#L338)):*
> ```Go
> // assertScanRowCounts performs assertions to prevent silently returning
> // incorrect results, e.g., if an index is corrupt. In the common case, the
> // number of fetched rows in an index join should be equal to the nubmer of
> ```

nit: `s/nubmer/number/`.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thank you!

> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148538#-:-OTT86rBBb8gI1H4lBtk:bro3093)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

mgartner committed:
sql: add fetch row count assertions in row-by-row index joiner

Assertions have been added to the join reader that ensure that an index
join fetches the expected number of rows.

Informs #135696

Release note: None

mgartner committed:
sql: add fetch row count assertions in row-by-row lookup joiner

Assertions have been added to the join reader that ensure that a lookup
join on key columns fetches the expected number of rows.

Fixes #135696

Release note: None

mgartner commented:
PTAL. I added assertions for row-by-row index joins and lookup joins where the lookup columns are keys.

yuzefovich reviewed (approved):
Thanks!

> Reviewed 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148538#-:-OTe91yWEQ-H_70e3u8z:b-ucfknh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner)



<!-- Sent from Reviewable.io -->
